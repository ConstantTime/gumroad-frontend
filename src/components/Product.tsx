import Typography from "@material-ui/core/Typography";
import React from "react";

const Product: React.FC = () => {
  return (
    <Typography variant="h3" component="h2">
      h1. Heading
    </Typography>
  );
};

export default Product;
