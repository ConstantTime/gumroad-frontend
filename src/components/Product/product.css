.product-heading {
  justify-content: center;
  margin-top: 20px;
}

.product-page {
  display: flex;
  flex-direction: column;
  width: 50%;
  margin: 0 auto;
}

.average-rating {
  align-items: center;
  justify-content: center;
  display: flex;
}

.rating-text {
  margin-right: 10px;
}

.add-rating {
  margin-top: 10px;
  align-items: center;
  justify-content: space-between;
  display: flex;
}

.product-rating {
  display: flex;
}

.separator {
  margin-top: 40px;
  margin-bottom: 40px;
  border-bottom: 2px solid;
}

.top3-reviews .MuiTypography-root {
  margin-bottom: 10px;
}
