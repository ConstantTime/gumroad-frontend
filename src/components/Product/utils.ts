import { IRating } from "../../api/responses/types";

export const averageRating = (ratings: IRating[]) => {
  if (ratings.length === 0) {
    return 0.0;
  }
  let avgRating = 0.0;
  ratings?.forEach((rating) => {
    avgRating = avgRating + rating.rating;
  });

  avgRating /= ratings?.length;

  return +avgRating.toFixed(2);
};
