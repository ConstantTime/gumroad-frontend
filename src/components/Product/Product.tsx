import Modal from "@material-ui/core/Modal";
import Typography from "@material-ui/core/Typography";
import Rating from "@material-ui/lab/Rating";
import axios from "axios";
import React, { useEffect, useState } from "react";
import { useParams } from "react-router-dom";
import { IProduct, IRating } from "../../api/responses/types";
import { ApiEndpoints, SERVER_URL } from "../../constants/Constants";
import AddRatingModal from "../AddRating/AddRatingModal";
import GumroadButton from "../Button/GumroadButton";
import Review from "../Review/Review";
import "./product.css";

const Product: React.FC = () => {
  const id = useParams<any>();
  const [product, setProduct] = useState<IProduct>();
  const [ratings, setRatings] = useState<IRating[]>([]);
  const [isModelOpen, showModal] = useState(false);

  const averageRating = () => {
    let avgRating = 0.0;
    ratings?.forEach((rating) => {
      avgRating = avgRating + rating.rating;
    });

    avgRating /= ratings?.length;

    return avgRating;
  };

  useEffect(() => {
    axios
      .get<IProduct>(
        SERVER_URL + ApiEndpoints.FETCH_PRODUCT_DETAILS(id.id as String),
        {
          method: "GET",
        }
      )
      .then((response) => {
        setProduct(response.data);
      })
      .catch((response) => {
        console.log("Error while fetching product details:", response);
      });

    axios
      .get<IRating[]>(
        SERVER_URL + ApiEndpoints.FETCH_TOP3_RATINGS(id.id as String),
        {
          method: "GET",
        }
      )
      .then((response) => {
        setRatings(response.data);
      })
      .catch((response) => {
        console.log("Error while fetching top 3 ratings:", response);
      });
  }, []);

  if (!product) {
    return <div>Error in loading product!!!</div>;
  }

  const handleClick = () => {
    showModal(true);
  };

  const handleClose = () => {
    showModal(false);
  };

  return (
    <div className="product-page">
      <Typography variant="h2" component="h2" align="center">
        <div className="product-heading">{product.name}</div>
      </Typography>
      <div className="add-rating">
        <div className="average-rating">
          <div className="rating-text">
            <Typography variant="h5" align="center">
              {averageRating()}
            </Typography>
          </div>
          <Rating
            name="half-rating"
            value={averageRating()}
            precision={0.5}
            readOnly
          />
        </div>

        <div className="add-rating-button">
          <GumroadButton onClick={handleClick}>Add Review</GumroadButton>
        </div>
      </div>

      <div className="separator"></div>

      <div className="top3-reviews">
        <Typography variant="h5" component="h2">
          Reviews
        </Typography>

        {ratings.map((rating) => (
          <Review rating={rating} />
        ))}
      </div>
      {isModelOpen && (
        <Modal
          open
          onClose={handleClose}
          aria-labelledby="simple-modal-title"
          aria-describedby="simple-modal-description"
        >
          <AddRatingModal />
        </Modal>
      )}
    </div>
  );
};

export default Product;
