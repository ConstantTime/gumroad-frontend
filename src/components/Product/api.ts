import axios from "axios";
import { IProduct, IRating } from "../../api/responses/types";
import { ApiEndpoints, SERVER_URL } from "../../constants/Constants";
import "./product.css";

export const fetchProductDetails = (
  id: any,
  setProduct: (data: IProduct | undefined) => void
) => {
  axios
    .get<IProduct>(
      SERVER_URL + ApiEndpoints.FETCH_PRODUCT_DETAILS(id.id as String),
      {
        method: "GET",
      }
    )
    .then((response) => {
      setProduct(response.data);
    })
    .catch((response) => {
      console.log("Error while fetching product details:", response);
    });
};

export const fetchTop3Ratings = (
  id: any,
  setRatings: (data: IRating[]) => void
) => {
  axios
    .get<IRating[]>(
      SERVER_URL + ApiEndpoints.FETCH_TOP3_RATINGS(id.id as String),
      {
        method: "GET",
      }
    )
    .then((response) => {
      setRatings(response.data);
    })
    .catch((response) => {
      console.log("Error while fetching top 3 ratings:", response);
    });
};
