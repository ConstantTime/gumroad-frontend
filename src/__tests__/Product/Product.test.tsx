import axios from "axios";
import { shallow } from "enzyme";
import Product from "../../components/Product/Product";

jest.mock("axios");
jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useParams: () => ({
    id: {
      id: "test",
    },
  }),
}));
axios.get = jest.fn().mockImplementation(() =>
  Promise.resolve({
    data: {
      productId: "1",
      name: "test",
      description: "description",
    },
  })
);

describe("Product", () => {
  it("should render the component", () => {
    const wrapper = shallow(<Product />);
    expect(wrapper).toMatchSnapshot();
  });

  it("should render the erroneous state of component", () => {
    const wrapper = shallow(<Product />);
    expect(wrapper).toMatchSnapshot();
  });
});
