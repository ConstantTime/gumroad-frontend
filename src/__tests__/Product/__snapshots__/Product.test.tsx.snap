// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product should render the component 1`] = `
<div>
  Error in loading product!!!
</div>
`;

exports[`Product should render the erroneous state of component 1`] = `
<div>
  Error in loading product!!!
</div>
`;
